solver/src/paraswap_solver/api.rs

```diff
@@ -0,0 +1,420 @@
+#![allow(dead_code)]
+
+use std::convert::TryFrom;
+
+use anyhow::{Context, Result};
+use ethcontract::{H160, U256};
+use reqwest::{Client, RequestBuilder, Url};
+use serde::{Deserialize, Serialize};
+use serde_json::{Map, Value};
+
+use model::u256_decimal;
+use web3::types::Bytes;
+
+const BASE_URL: &str = "https://apiv4.paraswap.io";
+
+/// Mockable implementation of the API for unit test
+#[cfg_attr(test, mockall::automock)]
+#[async_trait::async_trait]
+pub trait ParaswapApi {
+    async fn price(&self, query: PriceQuery) -> Result<PriceResponse>;
+    async fn transaction(
+        &self,
+        query: TransactionBuilderQuery,
+    ) -> Result<TransactionBuilderResponse>;
+}
+
+struct DefaultParaswapApi {
+    client: Client,
+}
+
+#[async_trait::async_trait]
+impl ParaswapApi for DefaultParaswapApi {
+    async fn price(&self, query: PriceQuery) -> Result<PriceResponse> {
+        reqwest::get(query.into_url())
+            .await
+            .context("PriceQuery failed")?
+            .json::<Value>()
+            .await
+            .context("PriceQuery result parsing failed")
+            .and_then(PriceResponse::try_from)
+    }
+    async fn transaction(
+        &self,
+        query: TransactionBuilderQuery,
+    ) -> Result<TransactionBuilderResponse> {
+        query
+            .into_request(&self.client)
+            .send()
+            .await
+            .context("TransactionBuilderQuery failed")?
+            .json()
+            .await
+            .context("TransactionBuilderQuery result parsing failed")
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum Side {
+    Buy,
+    Sell,
+}
+
+/// Paraswap price quote query parameters.
+#[derive(Clone, Debug)]
+pub struct PriceQuery {
+    /// source token address
+    pub from: H160,
+    /// destination token address
+    pub to: H160,
+    /// decimals of from token (according to API needed  to trade any token)
+    pub from_decimals: usize,
+    /// decimals of to token (according to API needed to trade any token)
+    pub to_decimals: usize,
+    /// amount of source token (in the smallest denomination, e.g. for ETH - 10**18)
+    pub amount: U256,
+    // Type of order
+    pub side: Side,
+}
+
+impl PriceQuery {
+    pub fn into_url(self) -> Url {
+        let mut url = Url::parse(BASE_URL)
+            .expect("invalid base url")
+            .join("v2/prices")
+            .expect("unexpectedly invalid URL segment");
+
+        let side = match self.side {
+            Side::Buy => "BUY",
+            Side::Sell => "SELL",
+        };
+
+        url.query_pairs_mut()
+            .append_pair("from", &format!("{:#x}", self.from))
+            .append_pair("to", &format!("{:#x}", self.to))
+            .append_pair("amount", &self.amount.to_string())
+            .append_pair("side", side)
+            .append_pair("network", "1");
```

Comment:
Do they support xDAI and Rinkeby? Might be nice to add support for it in the future (doesn't have to be this PR).